### PR TITLE
VideoCommon: flush vertex manager if components change

### PR DESCRIPTION
**Pull request by Tilka**
~~DON'T MERGE, this PR is only to get a FifoCI comparison.~~ I replaced the workaround with a proper fix.

degasus helped a lot with explaining and debugging this issue.

### CODE REVIEW

**Comment by degasus** [comment]
LGTM

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 1]
```diff
@@ -32,6 +32,7 @@ u32 position_matrix_index[3];
 typedef std::unordered_map<PortableVertexDeclaration, std::unique_ptr<NativeVertexFormat>> NativeVertexFormatMap;
 static NativeVertexFormatMap s_native_vertex_map;
 static NativeVertexFormat* s_current_vtx_fmt;
+u32 g_current_components;
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: Maybe move the comment from https://github.com/dolphin-emu/dolphin/pull/3218/files#diff-5ff1a45ad75bac713802503f8b4ef4f5L118 to here?

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 2]
```diff
@@ -185,9 +185,13 @@ int RunVertices(int vtx_attr_group, int primitive, int count, DataReader src, bo
...
 
 	// If the native vertex format changed, force a flush.
-	if (loader->m_native_vertex_format != s_current_vtx_fmt)
+	if (loader->m_native_vertex_format != s_current_vtx_fmt ||
+	    loader->m_native_components != g_current_components)
+	{
 		VertexManager::Flush();
+	}
 	s_current_vtx_fmt = loader->m_native_vertex_format;
+	g_current_components = loader->m_native_components;
```
Comment (Tilka) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: (This is the actual fix.)

**Comment by delroth** [comment]
Any reason to make shadergen depend on more global state vs. the previous way of passing the relevant state through args?

This doesn't seem to be justified and I see it as a regression.

**Comment by degasus** [comment]
@delroth This state could be passed by argument for SetShader(), but there is no (nice) way to do the same for VertexManager::Flush. Right now, only the vertex loader manager will know the last state, but almost every videocommon class might call VertexManager::Flush (which calls SetShader within the backends...). So we must still use a global. In my opinion, is cleaner to fetch the global where it's needed (shadergen), not within a completely different subproject (videocommon vs backends). If we want to pass the "state" by argument, we should think about a united "GPU state"...

**Comment by delroth** [comment]
Agreed. Thanks for explaining why the change is necessary (VertexManager::Flush).

**Pull request merged**